Add a Specified Flow Element to AirflowNetwork

jasondegraw opened the pull request:
Pull request overview
---------------------
 Add a specified flow element to Airflow Network.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jasondegraw committed:
Create NFP_AirflowNetwork_Specified_Flow_Element.md

shorowit commented:
@jasondegraw Could this object include both a flow rate and a reference to a schedule, so that it's user-specified but not necessarily constant? That would make it far more flexible/useful. And then it would be named `AirflowNetwork:MultiZone:SpecifiedFlowRate` (or similar) instead of `AirflowNetwork:MultiZone:ConstantFlowRate`.

jasondegraw commented:
@shorowit We're open to different naming, but I don't think we want to add a schedule here. Scheduling fits better at the linkage level, and some of that is already possible. We have a second new feature for this FY to generalize linkage scheduling, so after that feature goes in the functionality you're talking about should be there.

nrel-bot-3 commented:
@jasondegraw @lgentile it has been 28 days since this pull request was last updated.

jasondegraw committed:
Update NFP_AirflowNetwork_Specified_Flow_Element.md

lgu1234 commented on design/FY2021/NFP_AirflowNetwork_Specified_Flow_Element.md:
```diff
@@ -0,0 +1,80 @@
+      \key scfm
+      \key slpm
+      \default kg/s
+      \note Enter the air flow units for this element.
```
Units of airflow inputs in E+ are m3/s. Is it possible to be consistent with others?

jasondegraw replied:
@lgu1234 Yes, we can do that. How about changing the choices to be kg/s and m3/s?

mjwitte replied:
For m3/s there's always the question of the density basis: outdoor current, indoor current, or standard? Another choice field? I've been considering such a choice field for the simple airflow objects (ZoneVentilation:* and ZoneInfiltration:*). Or maybe in this case it's just assumed to be standard density to align with HVAC system flow rate inputs?

mjwitte replied:
I guess `scfm` was going to be one of those. Anyways, not sure if the new editor will support it, but for field N1, you can add
```
      \unitsBasedOnField A2
```
Or just have two N fields:
```
A2, Air Flow Input Basis
  /key MassFlowRate
  /key VolumeFlowRateStandardDensity
  /key VolumeFlowRateOutdoorDensity
  /key VolumeFlowRateZoneDensity
N1, Air Mass Flow Rate
  /units kg/s
N2, Air Volume Flow Rate
  /units m3/s
```

Just thinking out loud.

jasondegraw replied:
After review, I think we're going to go with kg/s and m3/s and we'll go with the `\unitsBasedOnField` approach.

prsh5175 committed:
Merge remote-tracking branch 'origin/develop' into afn-specified-flow

nrel-bot-3 commented:
@jasondegraw @lgentile it has been 28 days since this pull request was last updated.

prsh5175 committed:
Start implementation of specified flow element

prsh5175 committed:
Merge remote-tracking branch 'origin/develop' into afn-specified-flow

prsh5175 committed:
Initial calculate implementations

nrel-bot-3 commented:
@jasondegraw @lgentile it has been 28 days since this pull request was last updated.

Myoldmopar commented:
With the IO Freeze just days away, it looks like this one might not make it.  I'm going through the PR list to update the milestones for the ones that might not make it so that our remaining review time is focused.  Obviously if this one is still on target to make it in, just swap the milestone back and we'll keep it in the queue!  Thanks! :+1:

jasondegraw committed:
Merge branch 'develop' into afn-specified-flow

jasondegraw committed:
Add input, unit tests, and documentation

jasondegraw committed:
Add specified flow element into a test file

And also fix up a bunch of stuff. The flow rate in the specified
flow element is set to the average hourly value over the two run
periods to hopefully minimize the impact of the element on the
model results.

jasondegraw commented on src/EnergyPlus/AirflowNetwork/include/AirflowNetwork/Elements.hpp:
```diff
@@ -147,7 +149,9 @@ namespace AirflowNetwork {
-        REL      // Distribution system relief air
+        REL,     // Distribution system relief air
+        SMF,     // Specified mass flow component
+        SVF      // Specified volume flow component
```
The less said about why there are currently two enumerations, the better.

jasondegraw commented on src/EnergyPlus/AirflowNetwork/include/AirflowNetwork/Elements.hpp:
```diff
@@ -573,6 +577,62 @@ namespace AirflowNetwork {
+            return ComponentType::SVF;
+        }
+    };
+
```
We ended up adding two separate objects to avoid explicit logical checks as to whether the object is supposed to be fixed mass flow or fixed volume flow.

jasondegraw commented on src/EnergyPlus/AirflowNetwork/src/Elements.cpp:
```diff
@@ -3370,6 +3370,86 @@ namespace AirflowNetwork {
+
+        return 1;
+    }
+
```
These two functions are the main part of the addition.

jasondegraw commented on src/EnergyPlus/AirflowNetworkBalanceManager.cc:
```diff
@@ -1120,14 +1120,59 @@ namespace AirflowNetworkBalanceManager {
+                }
+            }
+        }
+
```
Read the input object into the appropriate object. This would be simpler and look a lot nicer if real constructors were the order of things, but alas, no.

jasondegraw commented on src/EnergyPlus/AirflowNetworkBalanceManager.cc:
```diff
@@ -4730,6 +4776,37 @@ namespace AirflowNetworkBalanceManager {
+            ++ii;
+            ++type_i;
+        }
+
```
Another thing that's probably best left alone. This code is necessary for a couple of places that do reverse lookups of objects. "Fortran polymorphism with array indices" would probably be the best way to describe it.

jasondegraw commented on tst/EnergyPlus/unit/AirflowNetworkSolver.unit.cc:
```diff
@@ -259,3 +259,96 @@ TEST_F(EnergyPlusFixture, AirflowNetwork_SolverTest_GenericCrack)
+    EXPECT_EQ(0.0, DF[0]);
+    EXPECT_EQ(0.0, DF[1]);
+    EXPECT_EQ(1, NF);
+}
```
And finally, a couple of unit tests that verify we always get back the same flow.

jasondegraw commented:
@Myoldmopar @mjwitte Ready for review. I did not add a new test IDF and changed up a current one a little, that's the file that is failing. I didn't want to add a new test file for this, but if that's the preference it is a pretty quick change. The NFP is a little out of date from where things up once we got in there and started working, I'll try to get that addressed later today but it should not hold up review.

jasondegraw commented:
A couple things to note on the example file changes. I took the average hourly value of flow through one of the walls from the the original example file and set the flow through that wall to always be that average value (0.06 kg/s). This was just over the two run periods in the file, so a full annual run may show more significant changes in energy use. When I have used these models I have often just taken out the run periods and done a full annual run, but I wasn't sure if that was just me. So it might be advisable to use the average value over the full year instead of what is there now.

lgu1234 commented:
I am OK with the code change. I also checked unit test and they worked as expected.

Then I run example file with the change for this new feature and found there are no corresponding outputs available. After adding two outputs:

ZN003:WALL001:AFN Linkage Node 1 to Node 2 Mass Flow Rate [kg/s](TimeStep)
ZN003:WALL001:AFN Linkage Node 2 to Node 1 Mass Flow Rate [kg/s](TimeStep)

Output shows 0.06 kg/s from  Node 1 to Node 2.

Then I change the flow value from + to -.

  AirflowNetwork:MultiZone:SpecifiedFlowRate,
    SF-1,                    !- Name
    -0.06,                    !- Air Flow Value
    kg/s;                    !- Air Flow Units

Output shows 0.06 kg/s from  Node 2 to Node 1 as expected.

I approve the PR with 2 minor comments:

1. Add output variables for this new feature in the example file
2. Add flow description in the object in IO ref.

lgu1234 commented on testfiles/AirflowNetwork_MultiZone_SmallOffice.idf:
```diff
@@ -1995,6 +1995,11 @@
+    SF-1,                    !- Name
+    0.06,                    !- Air Flow Value
+    kg/s;                    !- Air Flow Units
+
```
It is better to add output variables to demonstrate this new feature. For example:

ZN003:WALL001:AFN Linkage Node 1 to Node 2 Mass Flow Rate [kg/s](TimeStep)
ZN003:WALL001:AFN Linkage Node 2 to Node 1 Mass Flow Rate [kg/s](TimeStep)

jasondegraw replied:
@lgu1234 Oops, looks like I got confused about the file I was editing. My local tests had these in there, I'll fix that.

lgu1234 commented on doc/input-output-reference/src/overview/group-airflow-network.tex:
```diff
@@ -997,6 +1000,46 @@ \subsubsection{Inputs}\label{inputs-3-003}
+  0.0001,               !- Air flow rate
+  kg/s;                 !- Air flow units
+\end{lstlisting}
+
```
It is better to provide description for flow direction. The positive value show floe from thermal zone to outdoor....

jasondegraw replied:
@lgu1234 Good point, will add discussion of that as well.

jasondegraw committed:
Address comments from Gu

jasondegraw committed:
Update NFP_AirflowNetwork_Specified_Flow_Element.md

jasondegraw commented:
I've addressed the comments from @lgu1234 (thanks Gu!). I'll check in on CI later this evening and run the example again locally, but this should now be ready to go.

jasondegraw committed:
Fix a couple of small doc mistakes

jasondegraw committed:
Merge branch 'afn-specified-flow' of https://github.com/NREL/EnergyPlus into afn-specified-flow

jasondegraw committed:
One more small edit to the SFR docs

jasondegraw commented:
I think that's it for this one, I've cleaned up the docs a bit more and I'm not aware of anything else that is needed.

lgu1234 commented:
@Myoldmopar My minor comments were addressed. I satisfy the changes and approve the PR.

mitchute commented:
I just built and ran the tests here. Unit tests all ran to completion, and I ran the updated AirflowNetwork_MultiZone_SmallOffice file and did see that the `AFN Linkage Node 1 to Node 2 Mass Flow Rate` output variable is fixed at 0.06 kg/s as is specified by the new input object. I can't comment on the CI diffs, but based on these tests, the description in the NFP, and the conversation on the PR, this appears to be doing what it's supposed to.

I'll let someone else push the merge button.

Myoldmopar commented:
I built and ran regressions locally and it came back with only diffs in the one small office airflow network file.  This one is good to go in.  Thanks @jasondegraw and reviewers!
